src/overlay/OverlayManagerImpl.cpp

```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
 }
 
 void
-OverlayManagerImpl::storePeerList(std::vector<std::string> const& list,
-                                  bool setPreferred)
+OverlayManagerImpl::storePeerList(std::vector<PeerBareAddress> addresses,
+                                  bool setPreferred, bool startup)
 {
     auto typeUpgrade = setPreferred
                            ? PeerManager::TypeUpdate::SET_PREFERRED
                            : PeerManager::TypeUpdate::UPDATE_TO_OUTBOUND;
-
-    for (auto const& peerStr : list)
+    for (auto const& peer : addresses)
     {
-        try
+        if (setPreferred)
+        {
+            mConfigurationPreferredPeers.insert(peer);
+        }
+
+        if (startup)
         {
-            auto address = PeerBareAddress::resolve(peerStr, mApp);
-            getPeerManager().update(address, typeUpgrade,
+            getPeerManager().update(peer, typeUpgrade,
                                     PeerManager::BackOffUpdate::HARD_RESET);
         }
-        catch (std::runtime_error&)
+        else
         {
-            CLOG(ERROR, "Overlay") << "Unable to add peer '" << peerStr << "'";
+            // If address is present in the DB, `update` will ensure
+            // type is correctly updated. Otherwise, a new entry is created.
+            // Note that this won't downgrade promoted peers back to outbound.
+            getPeerManager().update(peer, typeUpgrade);
         }
     }
 }
 
 void
-OverlayManagerImpl::storeConfigPeers()
+OverlayManagerImpl::storeKnownAndPreferred(
+    std::vector<PeerBareAddress> known, std::vector<PeerBareAddress> preferred,
+    bool startup)
+{
+    storePeerList(std::move(known), false, startup);
+    storePeerList(std::move(preferred), true, startup);
+};
+
+void
+OverlayManagerImpl::storeConfigPeers(bool startup)
+{
+    if (startup)
+    {
+        // On startup, perform IP resolution synchronously on the main thread
+        auto known = resolvePeers(mApp.getConfig().KNOWN_PEERS);
+        auto preferred = resolvePeers(mApp.getConfig().PREFERRED_PEERS);
+        storeKnownAndPreferred(known, preferred, startup);
+    }
+    else
+    {
+        mApp.postOnBackgroundThread(
+            [this, startup]() {
+                auto known = resolvePeers(this->mApp.getConfig().KNOWN_PEERS);
+                auto preferred =
+                    resolvePeers(this->mApp.getConfig().PREFERRED_PEERS);
+
+                this->mApp.postOnMainThread(
```

Comment:
related to shutdown: can we actually post on main thread if overlay is shutting down?

Reply:
(Marking this as resolved as this is not an issue anymore with `std::future` approach)